photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,253 @@
+    http://www.cse.ust.hk/~psander/docs/gradsim.pdf
+
+"""
+import torch
+from typing import Optional, Union, Tuple, List
+
+import torch.nn.functional as F
+from torch.nn.modules.loss import _Loss
+
+from photosynthesis_metrics.utils import _adjust_dimensions, _validate_input
+
+
+def _prewitt_filter() -> torch.Tensor:
+    """Utility function that returns a normalized 3x3 prewitt kernel in X direction"""
+    # Normalize
+    kernel = torch.tensor([[-1., 0., 1.], [-1., 0., 1.], [-1., 0., 1.]]) / 3
+    return kernel
+
+
+def _gmsd(prediction: torch.Tensor, target: torch.Tensor,
+          reduction: Optional[str] = 'mean') -> torch.Tensor:
+    r"""Compute Gradient Magnitude Similarity Deviation
+    Both inputs supposed to be in range [0, 1] with RGB order.
+    Args:
+        prediction: Tensor of shape :math:`(N, C, H, W)` holding an distorted image.
+        target: Tensor of shape :math:`(N, C, H, W)` holding an target image
+        reduction: Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the sum of the output will be divided by the number of
+            elements in the output, ``'sum'``: the output will be summed.
+
+    Returns:
+        gmsd : Gradient Magnitude Similarity Deviation between given tensors.
+
+    References:
+        https://arxiv.org/pdf/1308.3052.pdf
+    """
+    # Constant for numerical stability
+    EPS: float = 1e-4
+
+    # Convert RGB image to YCbCr and take luminance: Y = 0.299 R + 0.587 G + 0.114 B
+    num_channels = prediction.size(1)
+    if num_channels == 3:
+        prediction = 0.299 * prediction[:, 0, :, :] + 0.587 * prediction[:, 1, :, :] + 0.114 * prediction[:, 2, :, :]
+        target = 0.299 * target[:, 0, :, :] + 0.587 * target[:, 1, :, :] + 0.114 * target[:, 2, :, :]
+
+        # Add channel dimension
+        prediction = prediction[:, None, :, :]
+        target = target[:, None, :, :]
+        
+    kernel = _prewitt_filter()
+    kernel_x = kernel.view(1, 1, 3, 3).to(prediction)
+    kernel_y = kernel.t().view(1, 1, 3, 3).to(prediction)
+
+    pred_grad_x = F.conv2d(prediction, kernel_x)
+    pred_grad_y = F.conv2d(prediction, kernel_y)
+    
+    trgt_grad_x = F.conv2d(target, kernel_x)
+    trgt_grad_y = F.conv2d(target, kernel_y)
+    
+    # Compute grad direction
+    pred_grad = torch.sqrt(pred_grad_x ** 2 + pred_grad_y ** 2)
+    trgt_grad = torch.sqrt(trgt_grad_x ** 2 + trgt_grad_y ** 2)
+    
+    # Compute GMS
+    gms = (2.0 * pred_grad * trgt_grad + EPS) / (pred_grad ** 2 + trgt_grad ** 2 + EPS)
+    mean_gms = torch.mean(gms, dim=[1, 2, 3], keepdims=True)
+
+    # Compute GMSD
+    gmsd = torch.pow(gms - mean_gms, 2).mean(dim=[1, 2, 3], keepdims=True).sqrt()
+    
+    if reduction == 'mean':
+        return gmsd.mean()
+    elif reduction == 'sum':
+        return gmsd.sum()
+    else:
+        return gmsd
```

Comment:
- Let's specify the dimension for the `mean` and `sum` operations. Therefore, we will have passive check that everything is calculated right and improve readability that we perform operation for batch dimension.
- Let's add check for specified `reduction` modes.
```suggestion
    if reduction == 'mean':
        return gmsd.mean(dim=0)
    elif reduction == 'sum':
        return gmsd.sum(dim=0)
    elif reduction != 'none':
        raise ValueError(f'Expected reduction modes are "mean"|"sum"|"none", got {reduction}')
    return gmsd
```